File: C++/Linked List/01_Reversing_linked_list.cpp
/*
Idea:
Iterating each element, and using temporary pointers, reversing the pointer direction
& returning the last element ( as head)
*/


//{ Driver Code Starts
#include <stdio.h>
#include <stdlib.h>
#include<iostream>
using namespace std;

/* Link list Node */
struct Node {
    int data;
    struct Node* next;
    Node(int x)
    {
        data = x;
        next = NULL;
    }
};


class Solution
{
public:
    //Function to reverse a linked list.
    struct Node* reverseList(struct Node* head)
    {
        // code here
        // return head of reversed list
        Node* now = head, * prev = NULL, * next;
        while (now != NULL)
        {
            next = now->next;
            now->next = prev;

            prev = now;
            now = next;
        }
        return prev;
    }

};



//{ Driver Code Starts.

void printList(struct Node* head)
{
    struct Node* temp = head;
    while (temp != NULL)
    {
        printf("%d ", temp->data);
        temp = temp->next;
    }
}

/* Driver program to test above function*/
int main()
{
    int T, n, l, firstdata;
    cin >> T;

    while (T--)
    {
        struct Node* head = NULL, * tail = NULL;

        cin >> n;

        if (n != 0)
        {
            cin >> firstdata;
            head = new Node(firstdata);
            tail = head;
        }

        for (int i = 1; i < n; i++)
        {
            cin >> l;
            tail->next = new Node(l);
            tail = tail->next;
        }

        Solution ob;
        head = ob.reverseList(head);

        printList(head);
        cout << endl;
    }
    return 0;
}


// } Driver Code Ends

File: Python/Linked List/01_Reversing_linked_list.py
"""
Idea:
Iterating each element, and using temporary pointers, reversing the pointer direction
& returning the last element ( as head)
"""




class Solution:
    # Function to reverse a linked list.
    def reverseList(self, head):
        now = head
        prev = None

        while now is not None:
            nxt = now.next
            now.next = prev

            prev = now
            now = nxt

        return prev


class Node:
    def __init__(self, val):
        self.data = val
        self.next = None


class Linked_List:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert(self, val):
        if self.head is None:
            self.head = Node(val)
            self.tail = self.head
        else:
            self.tail.next = Node(val)
            self.tail = self.tail.next


def printList(head):
    tmp = head
    while tmp:
        print(tmp.data, end=' ')
        tmp = tmp.next
    print()


if __name__ == '__main__':
    for i in range(int(input())):
        n = int(input())
        arr = [int(x) for x in input().split()]

        lis = Linked_List()
        for i in arr:
            lis.insert(i)

        newHead = Solution().reverseList(lis.head)
        printList(newHead)
